### DISCUSSION TITLE
Turn off the bank status LEDs just after changing the bank

### DISCUSSION THREAD

**Discussion by Nivolp1976** [tttapa/Control-Surface#524] [Q&A]
Hello, I would like to know if there is a way to turn off the bank status LEDs just after changing the bank or if the states can be updated without touching the potentiometers just after changing the bank. 

The problem is that with three sets of lights it becomes confusing that the lights of bank two remain on when moving to bank three and only update by moving the potentiometers. Thanks a lot!!

```cpp
  // You can request the state of the smart potentiometer:
 if (potentiometer1.getState() == CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer
    // value is too low to be active
    mcp.digitalWrite(LED_PIN4 , LOW);
    mcp.digitalWrite(LED_PIN3 , HIGH);
    mcp.digitalWrite(LED_PIN5 , LOW);
  } else if (potentiometer1.getState() == CCSmartPot::Higher) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
    mcp.digitalWrite(LED_PIN4 , LOW);
    mcp.digitalWrite(LED_PIN3 , LOW);
    mcp.digitalWrite(LED_PIN5 , HIGH);
  } else if (potentiometer1.getState() != CCSmartPot::Higher &&  CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
      mcp.digitalWrite(LED_PIN4 , HIGH);
      mcp.digitalWrite(LED_PIN3 , LOW);
      mcp.digitalWrite(LED_PIN5 , LOW);
  }

  // You can request the state of the smart potentiometer:
 if (potentiometer1.getState() == CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer
    // value is too low to be active
    mcp.digitalWrite(LED_PIN4 , LOW);
    mcp.digitalWrite(LED_PIN3 , HIGH);
    mcp.digitalWrite(LED_PIN5 , LOW);
  } else if (potentiometer1.getState() == CCSmartPot::Higher) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
    mcp.digitalWrite(LED_PIN4 , LOW);
    mcp.digitalWrite(LED_PIN3 , LOW);
    mcp.digitalWrite(LED_PIN5 , HIGH);
  } else if (potentiometer1.getState() != CCSmartPot::Higher &&  CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
      mcp.digitalWrite(LED_PIN4 , HIGH);
      mcp.digitalWrite(LED_PIN3 , LOW);
      mcp.digitalWrite(LED_PIN5 , LOW);
  }
```

**Comment by tttapa** [comment]
> I would like to know if there is a way to turn off the bank status LEDs just after changing the bank or if the states can be updated without touching the potentiometers just after changing the bank.

I'm not sure I understand what you mean by that, could you elaborate?

> `else if (potentiometer1.getState() != CCSmartPot::Higher &&  CCSmartPot::Lower)`

This probably doesn't do what you think it does, it should be `potentiometer1.getState() != CCSmartPot::Higher &&  potentiometer1.getState() != CCSmartPot::Lower`, but in this case, you could just make it an `else` clause instead of `else if`.

**Reply by Nivolp1976** [reply]
Sorry, I didn't explain it well. I'm just looking for the status LEDs of the smart potentiometers to update automatically, without the need to touch the potentiometers when changing banks. Thank you very much for responding, really.

**Reply by tttapa** [reply]
Ah I think I understand now, that's indeed an oversight. I've pushed a patch (9c6642836ceecc81f7dd3b5c7c2902a904b1cc5a), does that do what you expect?

**Reply by Nivolp1976** [reply]
This does what I expected perfectly, thank you so much for helping in so many ways. I share my absolute newbie code with everything working, several things I want to add are missing.


#include <Control_Surface.h> // Include the Control Surface library

#include <Adafruit_MCP23X17.h>


#include <U8g2lib.h>
#ifdef U8X8_HAVE_HW_I2C
#include <Wire.h>
#endif
U8G2_SSD1306_128X32_UNIVISION_F_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE);  


#define LED_PIN 0     // MCP23XXX pin LED is attached to
#define LED_PIN1 1     // MCP23XXX pin LED is attached to
#define LED_PIN2 2     // MCP23XXX pin LED is attached to
#define LED_PIN8 8     // MCP23XXX pin LED is attached to
#define LED_PIN9 9     // MCP23XXX pin LED is attached to
#define LED_PIN10 10     // MCP23XXX pin LED is attached to
#define LED_PIN3 3     // MCP23XXX pin LED is attached to
#define LED_PIN4 4     // MCP23XXX pin LED is attached to
#define LED_PIN5 5     // MCP23XXX pin LED is attached to
Adafruit_MCP23X17 mcp;

USBMIDI_Interface midi; // Instantiate a MIDI over USB interface.


Bank<3> bank {2};  // Four banks to change the channel of the potentiometer

// Selector to change banks
IncrementSelector<3> selector1 = {
  bank,       // Bank to manage
  2,          // push button pin
};

// Handy type alias
using  CCSmartPot = Bankable::CCSmartPotentiometer<3>;

// Instantiate a CCPotentiometer object
CCSmartPot potentiometer1 {
  {bank, BankType::CHANGE_ADDRESS},     // Bank configuration
  A3,                                   // Analog pin connected to potentiometer
  {MIDI_CC::General_Purpose_Controller_1 , CHANNEL_1}, // Channel volume of channel 1
};
CCSmartPot potentiometer2 {
{bank, BankType::CHANGE_ADDRESS},     // Bank configuration
  A0,                                   // Analog pin connected to potentiometer
  {MIDI_CC::General_Purpose_Controller_4, CHANNEL_1}, // Channel volume of channel 1
};

// Instantiate a PBPotentiometer object
PBPotentiometer potentiometer3 {
  A6,        // Analog pin connected to potentiometer
  CHANNEL_1, // MIDI Channel 1
};
// Instantiate a CCPotentiometer object
CCPotentiometer potentiometer4 {
  A1,                                   // Analog pin connected to potentiometer
  {MIDI_CC::Pan  , CHANNEL_1}, // Channel volume of channel 1
};

// Instantiate a Transposer that can transpose from one octave down to one
// octave up
Transposer<-5, +5> transposer(12);
 
// Instantiate a Selector to change the transposition
IncrementDecrementSelector<transposer.getNumberOfBanks()> selector {
  transposer,
  {9, 10},
  Wrap::Clamp,
  //Wrap::NoWrap,
};
 
Bankable::NoteButton buttons[] {
  {transposer, 8, MIDI_Notes::C(4)}, {transposer, 7, MIDI_Notes::D(4)},
  
};


void setup(){ {
  Control_Surface.begin(); // Initialize Control Surface
    u8g2.begin();
  //pinMode(LED_BUILTIN, OUTPUT);
Wire.begin ();
Wire.setClock (400000); 
}
if (!mcp.begin_I2C()) {
  //if (!mcp.begin_SPI(CS_PIN)) {
    Serial.println("Error.");
   // while (1);
  }
    //Led Config
    mcp.pinMode(LED_PIN, OUTPUT);
    mcp.pinMode(LED_PIN1, OUTPUT);
    mcp.pinMode(LED_PIN2, OUTPUT);
    mcp.pinMode(LED_PIN8, OUTPUT);
    mcp.pinMode(LED_PIN9, OUTPUT);
    mcp.pinMode(LED_PIN10, OUTPUT);
    mcp.pinMode(LED_PIN3, OUTPUT);
    mcp.pinMode(LED_PIN4, OUTPUT);
    mcp.pinMode(LED_PIN5, OUTPUT);
    //Led config
    mcp.digitalWrite(LED_PIN8 , LOW);
    mcp.digitalWrite(LED_PIN9 , LOW);
    mcp.digitalWrite(LED_PIN10 , LOW);
    mcp.digitalWrite(LED_PIN3 , LOW);
    mcp.digitalWrite(LED_PIN4 , LOW);
    mcp.digitalWrite(LED_PIN5 , LOW);
    
    }
    int count = millis(); 
void loop() {
  Control_Surface.loop(); // Update the Control Surface

  // You can request the state of the smart potentiometer:
 if (potentiometer1.getState() == CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer
    // value is too low to be active
    mcp.digitalWrite(LED_PIN4 , LOW);
    mcp.digitalWrite(LED_PIN3 , HIGH);
    mcp.digitalWrite(LED_PIN5 , LOW);
  } else if (potentiometer1.getState() == CCSmartPot::Higher) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
    mcp.digitalWrite(LED_PIN4 , LOW);
    mcp.digitalWrite(LED_PIN3 , LOW);
    mcp.digitalWrite(LED_PIN5 , HIGH);
  } else if (potentiometer1.getState() != CCSmartPot::Higher &&  CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
      mcp.digitalWrite(LED_PIN4 , HIGH);
      mcp.digitalWrite(LED_PIN3 , LOW);
      mcp.digitalWrite(LED_PIN5 , LOW);
  }

 // potentiometer2.update ();
  if (potentiometer2.getState() == CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer
    // value is too low to be active
    mcp.digitalWrite(LED_PIN10 , HIGH);
    mcp.digitalWrite(LED_PIN9 , LOW);
    mcp.digitalWrite(LED_PIN8 , LOW);
  } else if (potentiometer2.getState() == CCSmartPot::Higher) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
    mcp.digitalWrite(LED_PIN8 , HIGH);
    mcp.digitalWrite(LED_PIN9 , LOW);
    mcp.digitalWrite(LED_PIN10 , LOW);
  } else if (potentiometer2.getState() != CCSmartPot::Higher &&  CCSmartPot::Lower) {
    // Turn on an LED to indicate that the potentiometer 
    // value is too high to be active
      mcp.digitalWrite(LED_PIN9 , HIGH);
      mcp.digitalWrite(LED_PIN10 , LOW);
      mcp.digitalWrite(LED_PIN8 , LOW);
  }

 
  


//Setup leds bank sellection
 if (bank.getSelection() == 0) {mcp.digitalWrite(LED_PIN , HIGH);}
 if (bank.getSelection() != 0) {mcp.digitalWrite(LED_PIN , LOW);}
 
  if (bank.getSelection() == 1) {mcp.digitalWrite(LED_PIN1, HIGH);}
   if (bank.getSelection() != 1) {mcp.digitalWrite(LED_PIN1, LOW);}

   if (bank.getSelection() == 2) {mcp.digitalWrite(LED_PIN2, HIGH);}
   if (bank.getSelection() != 2) {mcp.digitalWrite(LED_PIN2, LOW);}

   Serial.print ("BANK____________");
   Serial.println (bank.getSelection());
   Serial.print ("ESTADO POT RNTRE 1 0 2____________");
   Serial.println (potentiometer1.getState());
   //potentiometer1.enable();
   Serial.print ("LECTURA POT entre 126 62 01____________");
   Serial.println (potentiometer1.getValue ());
   
 
   if (count < 100){
       count ++;
   } 
   else if (bank.getSelection() == 0) {u8g2.clearBuffer();          // clear the internal memory
   u8g2.setFont(u8g2_font_ncenB14_tr); // choose a suitable font
   u8g2.drawStr(0, 20, "BANK A");
   //u8g2.setCursor(70, 20);
   //u8g2.print(bank.getSelection());
   //u8g2.drawStr(0,10,"Hello World!");  // write something to the internal memory
   u8g2.sendBuffer();          // transfer internal memory to the display
   count = 0;
   }
   else if (bank.getSelection() == 1) {u8g2.clearBuffer();          // clear the internal memory
   u8g2.setFont(u8g2_font_ncenB14_tr); // choose a suitable font
   u8g2.drawStr(0, 20, "BANK B");
   //u8g2.setCursor(70, 20);
   //u8g2.print(bank.getSelection());
   //u8g2.drawStr(0,10,"Hello World!");  // write something to the internal memory
   u8g2.sendBuffer();          // transfer internal memory to the display
   count = 0;
   }
   else if (bank.getSelection() == 2) {u8g2.clearBuffer();          // clear the internal memory
   u8g2.setFont(u8g2_font_ncenB14_tr); // choose a suitable font
   u8g2.drawStr(0, 20, "BANK C");
   //u8g2.setCursor(70, 20);
   //u8g2.print(bank.getSelection());
   //u8g2.drawStr(0,10,"Hello World!");  // write something to the internal memory
   u8g2.sendBuffer();          // transfer internal memory to the display
   count = 0;
   }
}
